### ISSUE TITLE
Third Vive Tracker isn't Found

### ISSUE THREAD

**Issue by 200Tigersbloxed** [nicoco007/BeatSaberCustomAvatars#40]
Beat Saber Log File:
[_latest.log](https://github.com/nicoco007/BeatSaberCustomAvatars/files/4775946/_latest.log)

Custom Avatars Log File:
[_latest.log](https://github.com/nicoco007/BeatSaberCustomAvatars/files/4775947/_latest.log)

**Desktop**
- Operating system: Windows 10
- Headset and controllers: Rift CV1, Touch V1, 2 Base station 1.0's, 3 Vive Tracker Pro (2018/Blue logo), 

**Extra Information**

+ OpenVR-Advanced-Settings, OpenVR-Space-Calibrator, and xsOverlay set as start up items.
+ Calibrate on Start Enabled, First Person View Enabled, Resize Mode could be None or Arm Span, and Enable Floor Adjust is disabled.
+ Turned on in the following order: Oculus App/CV1, Left Controller, Right Controller, SteamVR, Startup Apps (see above), SteamVR Home, Left Foot Tracker, Right Foot Tracker, Pelvis Tracker, and then Beat Saber (being launched from SteamVR Home).
+ Trackers are set to the correct Part in the Vive Tracker Manager.

**Describe the bug**
When opening Beat Saber, avatar will calibrate, but the waist will not be found, and will follow the head's direction (turning your head to the right will move the waist to the right). Reading the logs, they will only output 2 of the trackers (the first two that were turned on), and not the third one.

Quote from Custom Avatars Log:
```
20 | [INFO @ 23:04:18 | Custom Avatars] Using device "VIVE Tracker Pro MV S/N LHR-8AB615DD" as LeftFoot
21 | [INFO @ 23:04:18 | Custom Avatars] Using device "VIVE Tracker Pro MV S/N LHR-5A80C2F6" as RightFoot
```

**Steps to reproduce**
Unknown, happens randomly. This is the second time it's happen. The first time it happened, I updated all the mods and libs with ModAssistant, and it just started working, but that won't solve it this time.

**Expected behavior**
Logs should output the Third Vive Tracker as the Pelvis, along with it's information, and avatar should be able to rotate waist without having to move the head.

**Comment by nicoco007** [comment]
Thanks for the detailed report!

Looking at the log, right above the "Using device <name> as <role>" lines, there's these lines:

```
459 | [INFO @ 23:04:18 | Custom Avatars] Found new input device "VIVE Tracker Pro MV S/N LHR-8AB615DD" with serial number "LHR-8AB615DD"
...
462 | [INFO @ 23:04:18 | Custom Avatars] Found new input device "VIVE Tracker Pro MV S/N LHR-5A80C2F6" with serial number "LHR-5A80C2F6"
```

Notice how only 2 devices called "VIVE Tracker Pro" are listed. These lines show the XR devices that Unity has access to, and a bit earlier in the log you can see the Versus mod showing similar information. Therefore, from what I can see at least, this seems to be some kind of miscommunication between SteamVR and Unity/Beat Saber causing the waist tracker to not discovered by Unity at all, not an issue specifically with Custom Avatars.

A few questions for you:
* If you turn the trackers on in a different order, is it always the same tracker that isn't discovered?
* Are you having similar issues in other games that use the trackers?
* If it's always the same tracker causing problems, could you:
  1. Make sure there are no firmware updates available for it
  2. Try resetting it by following these instructions from [the manual](https://fccid.io/NM82PYV300/User-Manual/User-Manual-3265250.pdf)
     > If you are having general issues with Vive Tracker, you can reset the hardware. Connect Vive Tracker to your computer using the supplied USB cable, and then press and hold the Power button for 10 seconds.

**Comment by 200Tigersbloxed** [comment]
Hey nicoco007, thanks for the fast response! Here's my answer to your questions.

1) No, say if I turned on Right Foot, waist, and then the Left Foot, the Left Foot won't be found. The Right Foor would be the Left Foot, and the Waist would be the Right Foot.
2) The only other game I use FBT with is VRChat, and it will always find the third tracker.
3) Firmware is up to date for both the Tracker and Base Stations, and is found by SteamVR and marked as Online when turned on.

**Comment by nicoco007** [comment]
Huh, that's weird. Here are my hypotheses on what might be going on:
* Since you say that updating your mods solved the issue in the past, it might be a conflict with another mod that's interfering with Unity for some reason. Does the problem still happen when running only Custom Avatars and its dependencies (DynamicOpenVR & BSML)?
* VRChat also uses Unity but is on 2018.4.20f1 while Beat Saber is on 2019.3.2f1; Unity's input system was revamped in 2019, so this might be an strange bug that appeared during that transition. If that's the case, there isn't much I can do about it.
* I'm doing something wrong when iterating devices (and so is Versus, apparently).

**Comment by 200Tigersbloxed** [comment]
Hey I'm back after doing what you suggested. I removed all unnecessary mods and tried again. No different result. It still was unable to find the third tracker.

**Comment by nicoco007** [comment]
I did some digging and it looks like Unity does some weird filtering of devices without telling anyone and doesn't have a way to bypass it. I tried with 11 devices plugged in (headset, 2 Vive controllers, 2 Index controllers, 3 Vive trackers, 2 base stations, and a gamepad); it only detects 6 to 8 of them, and it seems to depend on the order in which they're turned on. Unfortunately, I have no clue why this is happening and since the system Beat Saber uses is now considered "legacy" by Unity, I doubt this'll be fixed unless the Beat Saber devs decide to switch over to the new system.

**Comment by 200Tigersbloxed** [comment]
Maybe, instead of using Unity to find the devices, you could make a separate app, like a .Net C# app, that'll get the information for positioning and rotation, and whatnot, of the trackers and then have the mod read that data. It could be totally optional for people who want to use the mod without Full Body Tracking, and it wouldn't rely on Unity to get the tracker's information.

**Comment by 200Tigersbloxed** [comment]
Hey, I found an alternate solution! What I did, was before starting Oculus, I removed the batteries to the controllers, so they wouldn't be found, then I started everything up and when in Beat Saber, the hip tracker was found! But, 1 device has to be turned off, otherwise the right controller would not respond, not even to button presses, so I turned off 1 of my Oculus sensors and are using 2 sensors, to use the hip tracker. Not a good solution, but will work for now. Probably more proof that it is Unity messing up here.

**Comment by nicoco007** [comment]
Glad you got it to work (as hacky as it might be)! As for your previous suggestion, it's definitely possible but it'll take some work so it'll most likely have to wait until **after** this upcoming release since I've already been pushing it back for too long.

**Comment by 200Tigersbloxed** [comment]
Final comment for now. When I turned on Beat Saber today, all the trackers just started working again, without having to remove any devices. I don't know what I did, maybe it was switching 2 of my oculus sensor's usb ports, maybe it was plugging in my xbox controller and steam finding it, I don't know, but now it's just working.

**Comment by Kmagekris** [comment]
Hello. Just wanted to chime in I have had a similar issue, 4 basestations 2 index controllers, and 3 trackers. All are detected in Steam vr but beatsaber will only allow 4 to be detected at any time so if i turn a contoller off the other controller and 3 trackers are seen but if I turn on the controller, one of the trackers will dissapear.

**Comment by 200Tigersbloxed** [comment]
If you turn off 1 basestation, that should also work.

**Comment by Kmagekris** [comment]
Hi Tiger that did the trick! Thank you! :D

**Comment by nicoco007** [comment]
A fix for this will be included in the next release. If you're still having trouble once that's out, feel free to open another issue.
